### ISSUE TITLE
Write XMP Button should be disabled when action is in progress

### ISSUE THREAD

**Issue by koppor** [JabRef/jabref#8691]
We offer the functionality to write XMP to a PDF:

![grafik](https://user-images.githubusercontent.com/1366654/163879935-7e4b68b1-bbcd-4437-a75d-9c80fded1b32.png)

We tried to fix it at https://github.com/JabRef/jabref/pull/8658

However, the fix does not work:

<details>
<code>
com.google.common.util.concurrent.UncheckedExecutionException: java.lang.IllegalStateException: 1-based index not found: 8
	at org.jabref.merged.module@5.6.160/com.google.common.cache.LocalCache$Segment.get(Unknown Source)
	at org.jabref.merged.module@5.6.160/com.google.common.cache.LocalCache.get(Unknown Source)
	at org.jabref.merged.module@5.6.160/com.google.common.cache.LocalCache.getOrLoad(Unknown Source)
	at org.jabref.merged.module@5.6.160/com.google.common.cache.LocalCache$LocalLoadingCache.get(Unknown Source)
	at org.jabref.merged.module@5.6.160/com.google.common.cache.LocalCache$LocalLoadingCache.getUnchecked(Unknown Source)
	at org.jabref@5.6.160/org.jabref.logic.pdf.FileAnnotationCache.getFromCache(Unknown Source)
	at org.jabref@5.6.160/org.jabref.gui.entryeditor.fileannotationtab.FileAnnotationTabViewModel.lambda$reloadAnnotations$1(Unknown Source)
	at org.jabref.merged.module@5.6.160/com.sun.javafx.application.PlatformImpl.lambda$runLater$10(Unknown Source)
	at java.base/java.security.AccessController.doPrivileged(Unknown Source)
	at org.jabref.merged.module@5.6.160/com.sun.javafx.application.PlatformImpl.lambda$runLater$11(Unknown Source)
	at org.jabref.merged.module@5.6.160/com.sun.glass.ui.InvokeLaterDispatcher$Future.run(Unknown Source)
	at org.jabref.merged.module@5.6.160/com.sun.glass.ui.win.WinApplication._runLoop(Native Method)
	at org.jabref.merged.module@5.6.160/com.sun.glass.ui.win.WinApplication.lambda$runLoop$3(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
Caused by: java.lang.IllegalStateException: 1-based index not found: 8
	at org.jabref.merged.module@5.6.160/org.apache.pdfbox.pdmodel.PDPageTree.get(Unknown Source)
	at org.jabref.merged.module@5.6.160/org.apache.pdfbox.pdmodel.PDPageTree.get(Unknown Source)
	at org.jabref.merged.module@5.6.160/org.apache.pdfbox.pdmodel.PDPageTree.get(Unknown Source)
	at org.jabref.merged.module@5.6.160/org.apache.pdfbox.pdmodel.PDPageTree.get(Unknown Source)
	at org.jabref@5.6.160/org.jabref.logic.pdf.PdfAnnotationImporter.importAnnotations(Unknown Source)
	at org.jabref@5.6.160/org.jabref.logic.pdf.EntryAnnotationImporter.lambda$importAnnotationsFromFiles$2(Unknown Source)
	at java.base/java.util.Optional.ifPresent(Unknown Source)
	at org.jabref@5.6.160/org.jabref.logic.pdf.EntryAnnotationImporter.importAnnotationsFromFiles(Unknown Source)
	at org.jabref@5.6.160/org.jabref.logic.pdf.FileAnnotationCache$1.load(Unknown Source)
	at org.jabref@5.6.160/org.jabref.logic.pdf.FileAnnotationCache$1.load(Unknown Source)
	at org.jabref.merged.module@5.6.160/com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(Unknown Source)
	at org.jabref.merged.module@5.6.160/com.google.common.cache.LocalCache$Segment.loadSync(Unknown Source)
	at org.jabref.merged.module@5.6.160/com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(Unknown Source)
	... 14 more

</code>
</details>

Proposal: When clicking on the button

1. JabRef disalbes the button
2. JabRef does its work
3. JabRef enabled the button

This a) provides the user a feedback that JabRef is working and b) removes `synchronized` statements from the code.

**Comment by melmorsy** [comment]
Hi,
I'm interested in working on this issue, it's my first code contribution to open-source. I managed to get JabRef up and running from IntelliJ and I guess I found the button/action of interest (writeMetadataToPdf at LinkedFilesEditor, or am I at the wrong place?), however I'm not sure how to reproduce this issue, it would be great if someone could help please with providing steps to reproduce this issue from scratch as it's my first time to use JabRef.
Thanks

**Comment by ThiloteE** [comment]
Mhm I was the user that provided this error message to Oliver. I have not found out how to trigger the error message completely though. What I know:

1. Have an entry with bibliographic data
2. Have an pdf file attached to this entry
3. Press the button to write XMP metadata multiple times very fast in a row
4. ? -  I think there could be an additional condition. Maybe it has something to do with file annotations tab (which is trying to read annotations from the document).


**Additional Info:**

[This bug here](https://issues.apache.org/jira/browse/PDFBOX-4028?orderby=affectedVersion+ASC%2C+priority+DESC%2C+updated+DESC) is the underlying problem, which necessitates that saving to and reading from the PDF file should not happen at the same time. So in the context of the error message: I assume, apparently something still tries to read or write, while a write process is still ongoing, which is BAD.

**Comment by ThiloteE** [comment]
Thank you for your interest by the way! :-)

If you are here, you are at the right place:

![image](https://user-images.githubusercontent.com/73715071/164199022-9ee4d2df-1e4f-4f7a-b9e4-131c423fefc1.png)

**Comment by melmorsy** [comment]
Thank you @ThiloteE ! :) - I'm planning to push the proposed enhancement to disable the button once it's clicked then re-enable it once the write is done, will have a look at the contribution guidelines before pushing the change though, let me know please if you have any suggestions.

**Comment by the-star-sea** [comment]
hello.I am a new bee of open source project and I am interested in this issue.Could I have a try?

**Comment by ThiloteE** [comment]
Hello @the-star-sea I am thankful for your interest about working on JabRef.

It looks like melmorsy already is working on this, so unless melmorsy can't make it work or has lost interest, maybe it would be better to choose another issue.

If you are pretty new to open source programming, have a look at issues with the tag "[good first issues](https://github.com/JabRef/jabref/labels/good%20first%20issue)". If you are a student, the "[candidates for university  projects](https://github.com/JabRef/jabref/projects/9)" page offers some issues of varying difficulty and scope and that have been estimated to be compatible with university courses as well and often bring a larger feature to JabRef. Of course, feel free to tackle one of the other "normal" issues around, if you feel you have the knowledge or the willingness to take them on. There are many around!

Because there are so many issues, **prioritization** would boost JabRef's development. For example, if you have decided to work on a issue related to "search", have a look at multiple issues with the label "search" and only afterwards decide on which one you want to work on. What should be avoided: Having a look at the newest issues and blindly choose an issue from the first page. Thank you! There might be important issues on the first page, but there might also be very important issues further down the list. [Labels](https://github.com/JabRef/jabref/labels?page=1&sort=name-asc) will help you a lot. Furthermore, [bugs](https://github.com/JabRef/jabref/projects/5) are actively prioritized by JabRef maintainers and getting rid of high priority ones feels extra nice!

Having said all that, at the end of the day, of course, you are the one free to choose what to work on and what feature you prefer JabRef to have. Such is the nature of open source software.

Thank you!

**Comment by melmorsy** [comment]
Hi,
Apologies for the delayed reply, I have got a PR with a fix proposal however there's a suggestion there to use SimpleCommand that I didn't manage to figure out how it would work unfortunately: https://github.com/JabRef/jabref/pull/8728 - I'll add a comment to the PR to see how could we conclude this, or please let me know if you have any suggestions
